### PR TITLE
HttpWebRequest tests hardened & RequestStream tests disabled on Desktop

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/corefx/issues/19580

RequestStream tests aren't behaving correctly on Desktop and hang pretty often. Will follow up tomorrow and analyze the root cause.

Tested Innerloop in a loop 100 times
Tested Outerloop in a loop 100 times

thanks @safern for help!

### CODE REVIEW

**Comment by safern** [comment]
cc: @davidsh @stephentoub this fixes the hanging

**Comment by safern** [comment]
@dotnet-bot test Outerloop netfx Windows_NT Release x64 Build and Test

**Comment by safern** [comment]
@dotnet-bot test OuterLoop netfx Windows_NT Debug x64

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 1]
```diff
@@ -121,6 +121,7 @@ public void AllowReadStreamBuffering_SetTrueThenGet_ExpectTrue(Uri remoteServer)
             Assert.True(request.AllowReadStreamBuffering);
         }
 
+        [OuterLoop]
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Why was this moved to OuterLoop?

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 2]
```diff
@@ -135,14 +136,19 @@ public void AllowReadStreamBuffering_SetTrueThenGet_ExpectTrue(Uri remoteServer)
             }               
         }
 
-        [Theory, MemberData(nameof(EchoServers))]
-        public void ContentLength_SetAfterRequestSubmitted_ThrowsInvalidOperationException(Uri remoteServer)
+        [Fact]
+        public async Task ContentLength_SetAfterRequestSubmitted_ThrowsInvalidOperationException()
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: It's good to switch to a loopback, but why did this "need" to be done?  How does it relate to the deadlocks you mentioned?  If the original code caused deadlocks, that's a real problem.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 3]
```diff
@@ -178,14 +184,19 @@ public void ContentType_SetThenGetEmptyValue_ExpectNull(Uri remoteServer)
...
-                Assert.Throws<InvalidOperationException>(() => request.MaximumResponseHeadersLength = 255);
-            }
+                HttpWebRequest request = WebRequest.CreateHttp(uri);
+                Task<WebResponse> getResponse = request.GetResponseAsync();
+                await LoopbackServer.ReadRequestAndSendResponseAsync(server);
+                using (WebResponse response = await getResponse)
+                {
+                    Assert.Throws<InvalidOperationException>(() => request.MaximumResponseHeadersLength = 255);
+                }
+            });
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Same question

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 4]
```diff
@@ -221,14 +232,19 @@ public void MaximumAutomaticRedirections_SetThenGetOne_Success(Uri remoteServer)
...
-                Assert.Throws<InvalidOperationException>(() => request.ContinueTimeout = 255);
-            }
+                HttpWebRequest request = WebRequest.CreateHttp(uri);
+                Task<WebResponse> getResponse = request.GetResponseAsync();
+                await LoopbackServer.ReadRequestAndSendResponseAsync(server);
+                using (WebResponse response = await getResponse)
+                {
+                    Assert.Throws<InvalidOperationException>(() => request.ContinueTimeout = 255);
+                }
+            });
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Same question.

Also, all of these follow the same pattern.  They could all either:
a) just have a bunch of Assert.Throws in the same test to avoid making lots of additional requests, or
b) use a helper that contains the boilerplate

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 5]
```diff
@@ -301,7 +317,8 @@ public void TimeOut_SetThenGet_ValuesMatch(Uri remoteServer)
                 var sw = Stopwatch.StartNew();
                 WebException exception = Assert.Throws<WebException>(() =>
                 {
-                    var response = (HttpWebResponse)request.GetResponse();
+                    var response = request.GetResponse();
+                    response.Dispose();
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: I don't understand this change.  Isn't it GetResponse that's throwing the exception?  If so there won't ever be anything to Dispose.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 6]
```diff
@@ -331,14 +348,19 @@ public void UserAgent_SetThenGetWindows_ValuesMatch(Uri remoteServer)
...
-                Assert.Throws<InvalidOperationException>(() => request.Host = "localhost");
-            }
+                HttpWebRequest request = WebRequest.CreateHttp(uri);
+                Task<WebResponse> getResponse = request.GetResponseAsync();
+                await LoopbackServer.ReadRequestAndSendResponseAsync(server);
+                using (WebResponse response = await getResponse)
+                {
+                    Assert.Throws<InvalidOperationException>(() => request.Host = "localhost");
+                }
+            });
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Same question as earlier

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 7]
```diff
@@ -483,14 +505,19 @@ public void KeepAlive_CorrectConnectionHeaderSent(bool? keepAlive)
...
             {
-                Assert.Throws<InvalidOperationException>(() => request.AutomaticDecompression = DecompressionMethods.Deflate);
-            }
+                HttpWebRequest request = WebRequest.CreateHttp(uri);
+                Task<WebResponse> getResponse = request.GetResponseAsync();
+                await LoopbackServer.ReadRequestAndSendResponseAsync(server);
+                using (WebResponse response = await getResponse)
+                {
+                    Assert.Throws<InvalidOperationException>(() => request.AutomaticDecompression = DecompressionMethods.Deflate);
+                }
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Same question as earlier... will stop commenting on these

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs [hunk 8]
```diff
@@ -1039,7 +1071,8 @@ public void Range_Add_Success(Uri remoteServer)
         {
             HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
             request.UseDefaultCredentials = true;
-            await request.GetResponseAsync();
+            var response = await request.GetResponseAsync();
+            response.Dispose();
```
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: This is the first change in this PR where I understand why it needed to be done.

**Comment by stephentoub** [comment]
I'm going to merge this to unblock CI for the netfx leg, but please follow-up on my questions/comments subsequently.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 5 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: True as long as GetResponse throws WebException which is the intent of this test. If not, it should still do cleanup

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 1 above)
Comment (ViktorHofer) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: This is indeed wrong and should be moved back to InnerLoop again. Just verified it with loops

**Comment by ViktorHofer** [comment]
@stephentoub To remove the timeouts I had to debug these issues. I started in the HttpWebRequest class as there was a failing method in the log: https://ci.dot.net/job/dotnet_corefx/job/master/job/netfx_windows_nt_release_prtest/12/consoleFull#4967889112d31e50d-1517-49fc-92b3-2ca637122019

It seemed obvious to harden the test to make it more reliable, therefore I changed them to use a  LoopbackServer.

As the timeouts were still not going away, I realized that the RequestStream tests are also indefinitely hanging. So I had to disable the whole class (as @safern and I wanted to get clean CI again as serveral PRs were blocked). I will follow up with the exact issue in that class now.

File: src/System.Net.Requests/tests/HttpWebRequestTest.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.Net.Requests/tests/HttpWebRequestTest.cs]: Except now with your change you don't know whether the exception came from GetResponse or Dispose.  This should be reverted.
